pkg/resource/deploy/plan_executor.go

```diff
@@ -213,6 +195,74 @@ func (pe *planExecutor) Execute(callerCtx context.Context, opts Options, preview
 	return res
 }
 
+func (pe *planExecutor) performDeletes(ctx context.Context, opts Options) (bool, result.Result) {
+	defer func() {
+		// We're done here - signal completion so that the step executor knows to terminate.
+		pe.stepExec.SignalCompletion()
+	}()
+
+	logging.V(7).Infof("performDeletes(...): beginning")
+
+	deleteSteps, res := pe.stepGen.GenerateDeletes(opts.DestroyTargets)
+	if res != nil {
+		logging.V(7).Infof("performDeletes(...): generating deletes produced error result")
+		return false, res
+	}
+
+	deletes := pe.stepGen.ScheduleDeletes(deleteSteps)
+
+	// ScheduleDeletes gives us a list of lists of steps. Each list of steps can safely be executed
+	// in parallel, but each list must execute completes before the next list can safely begin
+	// executing.
+	//
+	// This is not "true" delete parallelism, since there may be resources that could safely begin
+	// deleting but we won't until the previous set of deletes fully completes. This approximation
+	// is conservative, but correct.
+	for _, antichain := range deletes {
+		logging.V(4).Infof("planExecutor.Execute(...): beginning delete antichain")
+		tok := pe.stepExec.ExecuteParallel(antichain)
+		tok.Wait(ctx)
+		logging.V(4).Infof("planExecutor.Execute(...): antichain complete")
+	}
+
+	// After executing targeted deletes, we may now have resources that depend on the resource that
+	// were deleted.  Go through and clean things up accordingly for them.
+
+	if len(opts.DestroyTargets) > 0 {
+		deletedUrns := make(map[resource.URN]bool)
```

Comment:
`refresh` has very similar (identical?) code--can we share it somehow?

Reply:
will see.

Reply:
Done.  Logic is now nearly entirely shared.